stablehlo/reference/InterpreterApi.cpp

```diff
@@ -0,0 +1,126 @@
+  llvm::StringRef funcName = currentFunction.getSymName();
+
+  if (auto probeOp = dyn_cast<stablehlo::interpreter::ProbeOp>(op)) {
+    auto input =
+        stablehlo::InterpreterValue(scope.findTensor(probeOp.getOperand()));
+    auto status = stablehlo::interpreter::evalProbeOp(
+        input, probeOp.getProbeId(), config->probeInstrumentationDir,
+        instrumentedTensors);
+    scope.add(probeOp.getResult(), input);
+    return wrapStatus(std::move(status), funcName, "interpreter.probe");
+  }
+
+  if (auto runParallelOp =
+          dyn_cast<stablehlo::interpreter::RunParallelOp>(op)) {
+    auto runtimeOperands = scope.find(runParallelOp.getInputs());
+    std::queue<StringAttr> infeed;
+    if (auto infeedAttr = runParallelOp.getInfeed())
+      for (auto &value : infeedAttr->getValue())
+        infeed.push(value.cast<FlatSymbolRefAttr>().getAttr());
+
+    SmallVector<SmallVector<StringAttr>> programs(
+        runParallelOp.getPrograms().size());
+    for (auto [i, replica] : llvm::enumerate(runParallelOp.getPrograms()))
+      for (auto &program : replica.cast<ArrayAttr>())
+        programs[i].push_back(program.cast<FlatSymbolRefAttr>().getAttr());
+
+    SymbolTable symbolTable{op.getParentOfType<ModuleOp>()};
+    auto results = stablehlo::interpreter::evalRunParallelOp(
+        runtimeOperands, infeed, programs, symbolTable);
+    scope.add(runParallelOp.getResults(), results);
+    return wrapStatus(llvm::Error::success(), funcName,
+                      "interpreter.run_parallel");
+  }
+
+  return handleOp(op, process, scope);
+}
+
+llvm::Error InterpreterFallback::handleOp(Operation &op, Process *process,
+                                          Scope &scope) {
+  return stablehlo::invalidArgument("Unsupported op: %s",
+                                    debugString(op).c_str());
+}
+
+llvm::ErrorOr<SmallVector<InterpreterValue>> runInterpreter(
+    ModuleOp module, ArrayRef<InterpreterValue> inputs,
+    const InterpreterConfiguration &config) {
+  auto numFuncs = 0;
+  bool hasMain = false;
+
+  assert(!config.mainFunction.empty() && "Main function name cannot be empty");
+  module.walk([&](func::FuncOp funcOp) {
+    if (funcOp.getSymName() == config.mainFunction) hasMain = true;
+    numFuncs++;
+  });
+
+  if (numFuncs > 1 && !hasMain)
+    llvm::report_fatal_error("Requested main function not found.");
+
+  if (!config.probeInstrumentationDir.empty()) {
+    llvm::SmallString<128> instrumentationMetadataFile(
+        config.probeInstrumentationDir);
+    llvm::sys::path::append(instrumentationMetadataFile,
+                            stablehlo::numpy::kInstrumentationMetadataFilename);
+    if (llvm::sys::fs::remove(instrumentationMetadataFile))
+      llvm::report_fatal_error(
+          "Failed to remove existing instrumentation metadata file.");
+  }
+
+  SmallVector<InterpreterValue> results;
+  config.fallback->setConfig(config);
+
+  auto walkResult = module.walk([&](func::FuncOp funcOp) {
```

Comment:
Per above comment, I don't think we need this walk, if the goal is to just find the main function for executing, lets get it above and just call eval here.